plugin/restore/mysql/mysql.go

```diff
@@ -680,6 +680,189 @@ func (r *Restore) GetSortedBinlogFilesMetaOnServer(ctx context.Context) ([]Binlo
+		"--base64-output=DECODE-ROWS",
+	}
+	cmd := exec.CommandContext(ctx, r.mysqlutil.GetPath(mysqlutil.MySQLBinlog), args...)
+	cmd.Stderr = os.Stderr
+	pr, err := cmd.StdoutPipe()
+	if err != nil {
+		return 0, err
+	}
+	s := bufio.NewScanner(pr)
+	if err := cmd.Start(); err != nil {
+		return 0, err
+	}
+	defer func() {
+		pr.Close()
+		err = cmd.Process.Kill()
+	}()
+
+	for s.Scan() {
+		line := s.Text()
+		eventTsParsed, found, err := parseBinlogEventTsInLine(line)
+		if err != nil {
+			return 0, fmt.Errorf("failed to parse binlog eventTs from mysqlbinlog output, error: %w", err)
+		}
+		if !found {
+			continue
+		}
+		eventTs = eventTsParsed
+		break
+	}
+
+	return eventTs, nil
+}
+
+// Use command like mysqlbinlog --start-datetime=targetTs binlog.000001 to parse the first binlog event position with timestamp equal or after targetTs.
+// TODO(dragonly): Add integration test.
+func (r *Restore) getBinlogEventPositionAtOrAfterTs(ctx context.Context, binlogFile BinlogFile, targetTs int64, isLastBinlogFile bool) (pos int64, err error) {
+	args := []string{
+		// Local binlog file path.
+		path.Join(r.binlogDir, binlogFile.Name),
+		// Tell mysqlbinlog to suppress the BINLOG statements for row events, which reduces the unneeded output.
+		"--base64-output=DECODE-ROWS",
+		// Instruct mysqlbinlog to start output only after encountering the first binlog event with timestamp equal or after targetTs.
+		"--start-datetime", formatDateTime(targetTs),
+	}
+	cmd := exec.CommandContext(ctx, r.mysqlutil.GetPath(mysqlutil.MySQLBinlog), args...)
+	cmd.Stderr = os.Stderr
+	pr, err := cmd.StdoutPipe()
+	if err != nil {
+		return 0, err
+	}
+	s := bufio.NewScanner(pr)
+	if err := cmd.Start(); err != nil {
+		return 0, err
+	}
+	defer func() {
+		pr.Close()
+		err = cmd.Process.Kill()
+	}()
+
+	var foundPos bool
+	for s.Scan() {
+		line := s.Text()
+		posParsed, found, err := parseBinlogEventPosInLine(line)
+		if err != nil {
+			return 0, fmt.Errorf("failed to parse binlog event start position from mysqlbinlog output, error: %w", err)
+		}
+		if !found {
+			continue
+		}
+		if posParsed == 4 {
+			// When invoking mysqlbinlog with --start-datetime, the first valid event will always be FORMAT_DESCRIPTION_EVENT which should be skipped.
+			continue
+		}
+		pos = posParsed
+		foundPos = true
+		break
+	}
+
+	if !foundPos {
+		// There's no binlog event in this binlog file with ts > targetTs, which means the next binlog file's first eventTs > targetTs.
+		// We should return the end position of the last binlog event, i.e, the size of the binlog file.
+		// However, if this is the last binlog file, which means the user wants to recover to a time in the future and we should return an error.
+		if isLastBinlogFile {
+			return 0, fmt.Errorf("the targetTs %d is after the last event ts of the latest binlog file %q", targetTs, binlogFile.Name)
+		}
+		return binlogFile.Size, nil
```

Comment:
Size looks quite confusing since that's implementation details. And Size might correspond to end position instead of the start position that we use canonically. Why not use pos?

Reply:
It's the end position of the last event, as well as the start position of the next event.
If we are to use position, it should be the first event position of the next binlog file, which is 4.